### PR TITLE
[clang] Implement CWG2428 "Deprecating a concept"

### PR DESCRIPTION
**Pull request by Endilll**
This patch allows attributes to be attached to C++20 concepts, implementing [CWG2428](https://cplusplus.github.io/CWG/issues/2428.html).

### CODE REVIEW

File: clang/www/cxx_dr_status.html [hunk 1]
```diff
@@ -10698,7 +10698,7 @@ <h2 id="cxxdr">C++ defect report implementation status</h2>
     <td><a href="https://cplusplus.github.io/CWG/issues/1815.html">1815</a></td>
     <td>CD4</td>
     <td>Lifetime extension in aggregate initialization</td>
-    <td class="unreleased" align="center">Clang 19</td>
+    <td class="full" align="center">Yes</td>
```
Comment (AaronBallman) [inline clang/www/cxx_dr_status.html:10701]: Unrelated changes snuck in.

File: clang/test/CXX/drs/cwg24xx.cpp [hunk 2]
```diff
@@ -45,6 +45,34 @@ void fallthrough(int n) {
 #endif
 }
 
+namespace cwg2428 { // cwg2428: 19
+#if __cplusplus >= 202002L
+template <typename>
+concept C [[deprecated]] = true; // #C
```
Comment (AaronBallman) [inline clang/test/CXX/drs/cwg24xx.cpp:51]: I'd like an additional test for rejecting:
```
template <typename>
[[deprecated]] concept C = true;
```
which isn't allowed per the grammar (though I find this slightly surprising and it may be an oversight from WG21).

It would also be good to have a test explicitly showing that we reject non-C++-style attributes.

**Review by AaronBallman — COMMENTED** [review]
Please be sure to add a release note about implementing the DR.

File: clang/www/cxx_dr_status.html (hunk 1 above)
Comment (Endilll) [reply clang/www/cxx_dr_status.html:10701]: This particular one is an unfortunate oversight from #87933, and will be fixed after #92113 is merged.
Other unrelated changes are by design since #90352, when we started to fetch CWG index from source that updates much more often that official publication.

Comment (AaronBallman) [reply clang/www/cxx_dr_status.html:10701]: It's usually better to apply the unrelated changes to the tree first, then base the patch on top of that; this way reviewers don't see unrelated changes in the patch.

Comment (Endilll) [reply clang/www/cxx_dr_status.html]: I updated the branch from `main`, now with #92113 changes. Unrelated changes are gone.

File: clang/test/CXX/drs/cwg24xx.cpp (hunk 2 above)
Comment (Endilll) [reply clang/test/CXX/drs/cwg24xx.cpp]: I added test with alternative attribute placing.

> It would also be good to have a test explicitly showing that we reject non-C++-style attributes.

It's not possible to write such a test, because we accept `deprecated` in more spellings than just C++11 one.

File: clang/test/CXX/drs/cwg24xx.cpp [hunk 3]
```diff
@@ -45,6 +45,38 @@ void fallthrough(int n) {
...
+  // expected-warning@-1 {{'C' is deprecated}}
+  //   expected-note@#cwg2428-C {{'C' has been explicitly marked deprecated here}}
+};
+
+void g(C auto a) {};
+// expected-warning@-1 {{'C' is deprecated}}
+//   expected-note@#cwg2428-C {{'C' has been explicitly marked deprecated here}}
+#endif
+} // namespace cwg2428
+
```
Comment (cor3ntin) [inline clang/test/CXX/drs/cwg24xx.cpp:79]: you are missing

```cpp=
template <C T>
void f() { 
  C auto foo T();
}
```

File: clang/test/CXX/drs/cwg24xx.cpp (hunk 2 above)
Comment (cor3ntin) [reply clang/test/CXX/drs/cwg24xx.cpp]: `concept C __attribute__((deprecated)) = true; ` should be ill-formed. that should be easy to test

Comment (cor3ntin) [reply clang/test/CXX/drs/cwg24xx.cpp]: As it turns out, GCC does support GNU-style attributes https://compiler-explorer.com/z/dccfdofKf
We should probably follow suite. Sorry for the confusion Vlad
(Which means you need to add tests + support for that)

File: clang/test/CXX/drs/cwg24xx.cpp (hunk 3 above)
Comment (Endilll) [reply clang/test/CXX/drs/cwg24xx.cpp:89]: Thank you for the test! It wasn't passing, so I added it and made it work.

File: clang/test/CXX/drs/cwg24xx.cpp (hunk 2 above)
Comment (Endilll) [reply clang/test/CXX/drs/cwg24xx.cpp]: Non-standard spellings of this attribute are out of scope of the Core issue, so I'd like to left them for a subsequent PR.

Comment (Endilll) [reply clang/test/CXX/drs/cwg24xx.cpp]: I added support for GNU and MSVC spellings.

File: clang/lib/Parse/ParseTemplate.cpp [hunk 4]
```diff
@@ -316,7 +318,8 @@ Parser::ParseConceptDefinition(const ParsedTemplateInfo &TemplateInfo,
   const IdentifierInfo *Id = Result.Identifier;
   SourceLocation IdLoc = Result.getBeginLoc();
 
-  DiagnoseAndSkipCXX11Attributes();
+  ParsedAttributes Attrs(AttrFactory);
+  MaybeParseAttributes(PAKM_GNU | PAKM_Declspec | PAKM_CXX11, Attrs);
```
Comment (cor3ntin) [inline clang/lib/Parse/ParseTemplate.cpp:322]: Did you check MSVC for the `PAKM_Declspec` ?

File: clang/lib/Sema/SemaChecking.cpp [hunk 5]
```diff
@@ -8092,6 +8092,18 @@ void Sema::checkCall(NamedDecl *FDecl, const FunctionProtoType *Proto,
     diagnoseArgDependentDiagnoseIfAttrs(FD, ThisArg, Args, Loc);
 }
 
+void Sema::CheckConstrainedAuto(TypeSourceInfo *TS) {
+  if (getLangOpts().CPlusPlus20) {
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaChecking.cpp:8096]: Does the check for `getLangOpts().CPlusPlus20` does anything here?

File: clang/lib/Sema/SemaChecking.cpp [hunk 6]
```diff
@@ -8092,6 +8092,18 @@ void Sema::checkCall(NamedDecl *FDecl, const FunctionProtoType *Proto,
...
+    if (const AutoType *AutoT = TS->getType()->getAs<AutoType>()) {
```
Comment (cor3ntin) [inline clang/lib/Sema/SemaChecking.cpp:8097]: I think this check should be kept on the call site and the function can take an `AutoType` as parameter

File: clang/lib/Parse/ParseTemplate.cpp (hunk 4 above)
Comment (Endilll) [reply clang/lib/Parse/ParseTemplate.cpp:322]: I checked that `deprecated` is supported by `__declspec` syntax https://learn.microsoft.com/en-us/cpp/cpp/declspec?view=msvc-170, but `__declspec` itself is not allowed anywhere in concept definition. I'll remove `PAKM_Declspec` then.

File: clang/lib/Sema/SemaChecking.cpp (hunk 6 above)
Comment (Endilll) [reply clang/lib/Sema/SemaChecking.cpp]: Fixed.

File: clang/lib/Sema/SemaChecking.cpp (hunk 5 above)
Comment (Endilll) [reply clang/lib/Sema/SemaChecking.cpp]: No, it doesn't. Removed.

**Review by erichkeane — APPROVED** [review]

**Review by cor3ntin — APPROVED** [review]
LGTM, thanks!

File: clang/lib/Parse/ParseTemplate.cpp [hunk 7]
```diff
@@ -167,9 +167,11 @@ Parser::DeclGroupPtrTy Parser::ParseTemplateDeclarationOrSpecialization(
                                   LastParamListWasEmpty);
 
   // Parse the actual template declaration.
-  if (Tok.is(tok::kw_concept))
-    return Actions.ConvertDeclToDeclGroup(
-        ParseConceptDefinition(TemplateInfo, DeclEnd));
+  if (Tok.is(tok::kw_concept)) {
+    Decl *ConceptDecl = ParseConceptDefinition(TemplateInfo, DeclEnd);
+    ParsingTemplateParams.complete(ConceptDecl);
```
Comment (shafik) [inline clang/lib/Parse/ParseTemplate.cpp:172]: Curious, why do we need this?

Comment (Endilll) [reply clang/lib/Parse/ParseTemplate.cpp:174]: I added a comment.

**Review by shafik — APPROVED** [review]
